File: src/main/java/org/neo/gomina/plugins/ssh/connector/SshClient.java
package org.neo.gomina.plugins.ssh.connector;

import com.jcraft.jsch.ChannelExec;
import com.jcraft.jsch.JSch;
import com.jcraft.jsch.JSchException;
import com.jcraft.jsch.Session;
import org.apache.commons.lang3.StringUtils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.neo.gomina.model.ssh.SshAuth;

import java.io.ByteArrayOutputStream;

public class SshClient {

    private static final Logger logger = LogManager.getLogger(SshClient.class);

    private JSch jsch = new JSch();
    private int connectTimeout = 3000;

    public Session getSession(String host, SshAuth auth) throws JSchException {
        Session session = jsch.getSession(auth.getUsername(), host, 22);
        if (StringUtils.isNotBlank(auth.getPassword())) {
            session.setPassword(auth.getPassword());
        }
        session.setConfig("StrictHostKeyChecking", "no");
        return session;
    }

    public void deploy(Session session, String applicationFolder, String version) throws Exception {
        String cmd = "sudo -u svc-ed-int /srv/ed/apps/" + applicationFolder + "/ops/release.sh " + version;
        String result = executeCommand(session, cmd);
    }

    public void run(Session session, String applicationFolder) throws Exception {
        String cmd = "sudo -u svc-ed-int /srv/ed/apps/" + applicationFolder + "/ops/run-all.sh";
        String result = executeCommand(session, cmd);
    }

    public void stop(Session session, String applicationFolder) throws Exception {
        String cmd = "sudo -u svc-ed-int /srv/ed/apps/" + applicationFolder + "/ops/stop-all.sh";
        String result = executeCommand(session, cmd);
    }

    public String whoami(Session session) throws Exception {
        return executeCommand(session, "whoami");
    }

    public String executeCommand(Session session, String cmd) throws Exception {
        logger.info("#CMD[" + session.getHost() + "]: " + cmd);
        long start = System.nanoTime();
        ChannelExec channel = (ChannelExec)session.openChannel("exec");
        channel.setPty(true);
        channel.setCommand(cmd);
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        channel.setOutputStream(baos);
        channel.connect();
        while (!channel.isClosed()) {
            Thread.sleep(2);
        }
        String res = new String(baos.toByteArray());
        logger.info("#RES: " + res + " in(" + (System.nanoTime() - start) + ")");
        return res;
    }

}

File: src/main/java/org/neo/gomina/integration/ssh/SshClient.kt
package org.neo.gomina.integration.ssh

import com.jcraft.jsch.ChannelExec
import com.jcraft.jsch.JSch
import com.jcraft.jsch.Session
import org.apache.commons.lang3.StringUtils
import org.apache.logging.log4j.LogManager
import java.io.ByteArrayOutputStream

private val logger = LogManager.getLogger(Session::class.java)

fun Session.sudo(user: String?, cmd: String) = sudo(user, listOf(cmd))

fun Session.sudo(user: String?, cmds: List<String>): String {
    val sudoPrefix = if (user?.isNotBlank() == true) "sudo -u $user " else ""
    return execute(cmds.map { sudoPrefix + it })
}

fun Session.execute(cmd: String) = execute(listOf(cmd))

fun Session.execute(cmds: List<String>): String {
    val cmd = cmds.joinToString(separator = " ; \n")
    logger.debug("#CMD[${this.host}]: $cmd")
    val start = System.nanoTime()
    val channel = this.openChannel("exec") as ChannelExec
    channel.setPty(true)
    channel.setCommand(cmd)
    val baos = ByteArrayOutputStream()
    channel.outputStream = baos
    channel.connect()
    while (!channel.isClosed) {
        Thread.sleep(2)
    }
    val res = String(baos.toByteArray())
    logger.debug("#RES: $res in ${(System.nanoTime() - start)} nano")
    return res
}

class SshClient {

    private val jsch = JSch()
    private val connectTimeout = 3000

    fun getSession(host: String, auth: SshAuth): Session {
        val session = jsch.getSession(auth.username, host, 22)
        if (StringUtils.isNotBlank(auth.password)) {
            session.setPassword(auth.password)
        }
        session.setConfig("StrictHostKeyChecking", "no")
        return session
    }

    companion object {
        private val logger = LogManager.getLogger(SshClient::class.java)
    }

}